xenstore-srv: implement buffered read/write

LKomaryanskiy opened the pull request:
Current implementation has a few issues such as:
1) In the case of the response buffer is overflowed due to malicious DumU (e.g. performs write operations but does not read responses), the response buffer will be corrupted because of the wrong condition. On the other hand, with the correct condition, overflowing response buffer will lead to infinite loop with trying to write zero-length data;
2) Fixed missing memory barriers in read/write operations;
3) Fixed reading/writing buffers of more than 1KB (size of ring buffers), because according to xenstore specification, it should be able to operate with buffers up to 4KB;

The PR fixes these issues.

firscity commented on xen-dom-mgmt/include/domain.h:
```diff
@@ -101,26 +103,14 @@ struct xen_domain_console {
 
 	/* TODO: domains can have more than one console */
 	struct xen_domain_console console;
-	struct k_sem xb_sem;
```
Better to make separation as one more commit

LKomaryanskiy replied:
Ok, thank you for your close review. I`ll separate changes to the several commits (where possible).

firscity commented on xen-dom-mgmt/src/xen-dom-mgmt.c:
```diff
@@ -106,10 +106,10 @@ static int allocate_domain_evtchns(struct xen_domain *domain)
 		return rc;
 	}
-	domain->xenstore_evtchn = rc;
+	domain->xenstore.xenstore_evtchn = rc;
```
(not your problem) Better to also rename it to smth like `remote_evtchn`

LKomaryanskiy replied:
Ok, I'll change naming.

firscity commented on xenstore-srv/include/xenstore_srv.h:
```diff
@@ -7,6 +7,32 @@
+	struct k_thread xenstore_thrd;
+	atomic_t xenstore_thrd_stop;
+	evtchn_port_t xenstore_evtchn;
+	evtchn_port_t local_xenstore_evtchn;
```
now it can be just `local_evtchn`

LKomaryanskiy replied:
I will change the naming. Also, I think all fields with the prefix "xenstore" can be renamed too (e.g. xenstore_thrd_stop, xenstore_thrd).

firscity commented on xenstore-srv/include/xenstore_srv.h:
```diff
@@ -7,6 +7,32 @@
+	char *buffer;
+};
+struct xenstore {
+	sys_dlist_t out_list;
```
may be a `sys_slist_t` probably (single-linked)

LKomaryanskiy replied:
I'll change it.

firscity commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -11,7 +11,6 @@
 #include <zephyr/init.h>
 #include <zephyr/xen/events.h>
 #include <zephyr/xen/public/hvm/params.h>
-#include <zephyr/xen/public/io/xs_wire.h>
```
Spurious change?

LKomaryanskiy replied:
No, I moved this include to [xenstore_srv.h](https://github.com/xen-troops/zephyr-xenlib/pull/42/files#diff-213cefb438cd570b950865c824564c26fb37e0daeae4b6247ce2e706df2da6df), so this is redundant include.

lorc reviewed (commented):
There are lots of the things are going in one commit. The very commit description, which now consists of 3 paragraphs hints that you can split it into at least three different commits :)

Please do this, it will be more manageable and easier to review

LKomaryanskiy commented on xenstore-srv/include/xenstore_srv.h:
```diff
@@ -7,6 +7,32 @@
+#include <zephyr/xen/public/io/xs_wire.h>
+
+struct buffered_data {
+	sys_dlist_t list;
```
Sure, single linked list will be more suitable in this case. I'll change it.

LKomaryanskiy commented:
@lorc thank you for your feedback. I'll try to separate changes (where possible) to the several commits. Currently, I mark this PR as a draft in order to other reviewers do not waste time on reviewing this version of PR.

LKomaryanskiy commented:
@lorc @firscity I separated changes into 3 commits and made proposed changes (use single-linked list, rename some fields in xenstore structure, etc). However, the PR fixes more issues than the number of patches, some changes can't be separated into isolated patches (e.g. adding memory barriers).
Also, I run checkpatch script locally and it doesn't print any errors, but CI has some error:
> No typos will be found - file '/var/lib/buildbot/hagrid/spelling.txt': No such file or directory

@lorc could you please check it?

LKomaryanskiy commented:
@lorc thank you. Fixed CI warning in commit message:
> WARNING:COMMIT_LOG_LONG_LINE: Possible unwrapped commit description (prefer a maximum 75 chars per line)

lorc commented:
regarding `ENOSYS` - could you please confirm that it is a part of the XenStore protocol?

LKomaryanskiy commented:
@lorc sure. As I see, in case of passing wrong message type, original xenstore sends [ENOSYS](https://github.com/xen-project/xen/blob/master/tools/xenstore/xenstored_core.c#L2017)

LKomaryanskiy commented:
Corrected description in commit which fixes read message functionality with a more accurate description of the main issue and intention of proposed changes.

lorc reviewed (commented):
Thank you for patches. It is better indeed that was before. But there are some comments.

1. Please describe how this new mechanism should work in high level. This will allow reviewers to validate your code against intended behaviour. Right now I don't fully understand what your idea is, so I can't say for sure if you code is doing what it is supposed to do.
2. Use appropriate types where possible. For example, `size_t` is the only correct type for sizing objects in memory. 
3. There is still space for Denial of Service attacks. As I can see, client may send requests indefinitely, without even reading responses. This will make backlog of unsent packets grow arbitrary big, eating all available heap memory. We need to impose some limits of unsent messages.

LKomaryanskiy reviewed (commented):
@lorc first of all, thank you for your close review, it really helps to do code better:)
> Please describe how this new mechanism should work in high ...

Ok, I will add a high-level description in commit messages.
> Use appropriate types where possible...

Mostly I use uint32_t (or analog type) for saving offset in buffer or passing length. In both cases, these values can't be more than 4096. Is it still worth using `size_t` type for these purposes? 

> here is still space for Denial of Service attacks...

Hm, I have a few options on how to fix it:
1. I can limit the number of available buffers, that can be allocated for client domain. For example, I can limit it to 10 buffers and add comment for this constant like:
/* This value is chosen randomly to prevent Denial of Service attacks. */ (or something like this)
And after we got this limit, we can say that client has some issues (for example some thread, that should read responces is hanged-up or even worse - this is a malicious client) so we can close xenstore connections for it.
2. I can try to implement a mechanism like in the original xenstore. As I see, they use timeouts for buffers and when it expires, they remove it. But I didn't investigate it close and it's one more question, which timeout would be OK.

As for me, the first option is better, because it's easier to implement and I think more reliable. Could you tell me, what do you think about this?

lorc commented:
>  Is it still worth using `size_t` type for these purposes?

Yes. In general it is good to use appropriate types whenever possible. Among other things, this includes `bool` for boolean values, `size_t` for sizes of objects in memory, `uintptr_t` for integer pointer representation and so on. In your particular case, most instances of `size`/`len` will be stored in CPU registers, which are 64 bits wide anyways. Yes, you can spare 4 bytes for `len` in your structures, but this is not worth it. There is a reason why all memory manipulation functions accept `size_t` and not some other type. 

> As for me, the first option is better, because it's easier to implement and I think more reliable. Could you tell me, what do you think about this?

I agree with you, the first option should be fine for our use case.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -255,40 +256,19 @@ static size_t get_input_offset(XENSTORE_RING_IDX cons, XENSTORE_RING_IDX prod,
-static size_t get_output_offset(XENSTORE_RING_IDX cons, XENSTORE_RING_IDX prod,
-				size_t *len)
+static uint32_t get_output_offset(XENSTORE_RING_IDX cons, XENSTORE_RING_IDX prod,
+				uint32_t *len)
```
In any case it will use exactly one CPU register, which is 64 bits wide. So you are not saving space there. On other hand, `size_t` is the correct type store to object sizes.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -243,12 +243,12 @@ static bool check_indexes(XENSTORE_RING_IDX cons, XENSTORE_RING_IDX prod)
 {
-	size_t delta = prod - cons;
-	*len = XENSTORE_RING_SIZE - MASK_XENSTORE_IDX(cons);
+	uint32_t delta = prod - cons;
```
Well, I believe there are helper macros to deal with ring buffers somewhere in xen headers.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -314,22 +294,131 @@ static size_t read_xb(struct xenstore *xenstore, uint8_t *data, uint32_t len)
+static int write_buf(struct xenstore *xenstore, const void *data, unsigned int len)
 {
+	uint32_t avail;
+	void *dest;
```
Ah, sorry, I thought that I've seen something like `dest +=`, but apparently I am wrong.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -1232,18 +1214,122 @@ const struct message_handle message_handle_list[XS_TYPE_COUNT] = { [XS_CONTROL]
+
+	if (!xenstore->in) {
+		xenstore->in = k_malloc(sizeof(*xenstore->in));
+		/* In case of allocation error, try to allocate it next time */
```
Yeah, I believe it will be better

LKomaryanskiy commented:
@lorc thank you again for your close review.
Please, take a look at a new version of PR with fixes.

In the new version, I made the following changes:
1) Change message_handle signature to pass as argument xenstore structure, instead of xen_domain as it was. Also, changed parameter types for the handlers from xen_doamin to xenstore structure;
2) Changed buffered_data structure: removed wr_hdr and hdr  fields. Now header is stored in the buffer. Also added comments with descriptions for fields;
3) Added limit of output buffers number to prevent Denial of Service attacks.
4) Added more details in the commit message about high-level design, the main idea of proposed changes, and why it's better than we have now;
5) Changed uint_32 to size_t in places, where we should use it;
6) Added error messages when we don't have enough memory to allocate buffer for output buffer;
7) Added error handling in case we're not able to allocate memory for the input buffer. It looks critical for Dom0, so I think it's better to invalidate client domain in this case, to save Dom0 workability;
8) Added parameters to invalidate_client function in order to give more information about the reason why we invalidate the client domain.

Also, I rechecked the overflowing mechanism in indexes and found only one place, where it was [incorrect](https://github.com/xen-troops/zephyr-xenlib/pull/42/files#diff-4375263f6665a9a61d50351a4ba33fdb9b8b85cad71b42a9f75ed0c4c8502705R358)

But I fixed this by changing condition from "bigger to "not equal". It would work correctly because before some operation we're anyway checking the indexes. The only one that we have to know to process ring_read is that cons index is not equal to prod.

The other places work properly, and my assumption about the wrong calculation was incorrect, I apologize for that. Let me prove this with the following example:
Let's use the same numbers, that I used before:
prod = 5
cons = 4294967292 (0xFFFFFFFC)
(prod - cons) would give as proper value 9.
That happens because we're subtracting cons, so this value is expectantly is wrapping-around:
0x0-(0xFFFFFFFC) = 4
0x0-(0xFFFFFFFD) = 3
0x0-(0xFFFFFFFE) = 2
0x0-(0xFFFFFFFF) = 1

I got a little bit confused when wrote my previous [assumption](https://github.com/xen-troops/zephyr-xenlib/pull/42#discussion_r1197929116) because I forget that we're not summing operands, we're subtracting them.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -314,22 +300,127 @@ static size_t read_xb(struct xenstore *xenstore, uint8_t *data, uint32_t len)
+		len = avail;
+	}
+
+	memcpy(dest, data, len);
```
I can't see how ring wrap around is handled in this function. In the worst case you'll write only one byte even if ring completely empty.

LKomaryanskiy replied:
Yes in this iteration we will write data to the end of the ring buffer, but `ring_write` is called from `handle_output`, which is called in the thread loop. So, the remaining data will be written in the next iteration.

lorc replied:
Yes, but you will send notify event. I am no sure that protocol permits this.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -1237,18 +1219,135 @@ const struct message_handle message_handle_list[XS_TYPE_COUNT] = { [XS_CONTROL]
+		len = avail;
+	}
+
+	memcpy(data, src, len);
```
I can't see how ring wrap around is handled in this function. In the worst case you'll read only one byte even if ring completely full.

LKomaryanskiy replied:
Yes in this iteration we will read data to the end of the ring buffer, but ring_read is called from handle_input, which is called in the thread loop. So, the remaining data will be read in the next iteration.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -1237,18 +1219,135 @@ const struct message_handle message_handle_list[XS_TYPE_COUNT] = { [XS_CONTROL]
+	dmb();
+	intf->req_cons += len;
+
+	notify_evtchn(xenstore->local_evtchn);
```
I am not sure if you supposed to send event if you didn't read the whole input ring. Probably not.

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -314,22 +294,131 @@ static size_t read_xb(struct xenstore *xenstore, uint8_t *data, uint32_t len)
+	struct xenstore_domain_interface *intf = xenstore->domint;
+
+	return !sys_slist_is_empty(&xenstore->out_list) &&
+		((intf->rsp_prod - intf->rsp_cons) < XENSTORE_RING_SIZE);
```
UPD: answered [here](https://github.com/xen-troops/zephyr-xenlib/pull/42#issuecomment-1559017744)

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -314,22 +300,127 @@ static size_t read_xb(struct xenstore *xenstore, uint8_t *data, uint32_t len)
+		return;
+	}
+
+	sys_slist_remove(&xenstore->out_list, NULL, &out->node);
```
Hw, this assert will fire here only if `(out->used > out->total_size)` because if less it's ok and we will return before this assert. But scenario when `(out->used > out->total_size)`, I think it never gonna happen because we pass to the second `ring_write` call `(out->total_size - out->used)` and we can't write more than total_size. Could you write some examples, even if it's unlikely when this can happen?

lorc replied:
Sole purpose of `ASSERT` is  to assert an invariant, i.e. something that is always true. So yes, we expect that in correctly written code, when it reaches this point, next statement is always true:  `out->used == out->total_size`. But, program logic that leads to this assertion is quite complex and is spread across several functions.  Someone in the future may change some bit of this logic and things may break. This is why I am proposing to add `ASSERT` there.

LKomaryanskiy replied:
Thank you for your explanation. Added.

LKomaryanskiy commented on xenstore-srv/include/xenstore_srv.h:
```diff
@@ -7,7 +7,23 @@
+	/* The total size of message header and payload */
+	size_t total_size;
+	/* Buffer with header and payload */
+	char *buffer;
```
Ok, I will fix that. Thank you:)

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -258,37 +265,16 @@ static size_t get_input_offset(XENSTORE_RING_IDX cons, XENSTORE_RING_IDX prod,
 {
-	size_t delta = XENSTORE_RING_SIZE - cons + prod;
-	*len = XENSTORE_RING_SIZE - MASK_XENSTORE_IDX(prod);
+	size_t free_space = XENSTORE_RING_SIZE - (prod - cons);
```
Ah, fair enough. Thanks for the clarification.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -314,22 +300,127 @@ static size_t read_xb(struct xenstore *xenstore, uint8_t *data, uint32_t len)
+	size_t total_size = len + sizeof(header);
+
+	if (xenstore->used_out_bufs > XENSTORE_MAX_OUT_BUF_NUM) {
+		invalidate_client(xenstore,
```
Yes, I do understand, that it is not easy to do. This is why I proposed to leave it as is and just add some comment that states that it can be done better somewhere in the future.

LKomaryanskiy replied:
Added comment.

LKomaryanskiy reviewed (commented):
The new PR has the following minor changes compared with the previous version:
1) According to MISRA-C standard, changed type of buffer from `char*` to `uint8_t*`;
2) Removed parentheses from define;
3) Added `__ASSERT_NO_MSG(out->used == out->total_size)` to `handle_output`, to prevent breaking in case of further code changing;
4) Added "TODO" comment to `send_reply_sz` with description to act more "softly" in case of number of available buffers is over.

LKomaryanskiy commented:
@firscity thank you for your close review! Added @firscity RB

lorc reviewed (approved):
`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

LKomaryanskiy commented:
Added @lorc RB, thank you!
